virtool/account/data.py

```diff
@@ -0,0 +1,364 @@
+        :param data: the update to the account
+        :return: the user account
+        """
+        update = {}
+
+        password = data.password
+        old_password = data.old_password
+        email = data.email
+
+        if password is not None:
+
+            if not await validate_credentials(self._db, user_id, old_password or ""):
+                raise ResourceError("Invalid credentials")
+
+            update = virtool.account.db.compose_password_update(password)
+
+        if email is not None:
+            update["email"] = email
+
+        if update:
+            document = await self._db.users.find_one_and_update(
+                {"_id": user_id},
+                {"$set": update},
+                projection=virtool.account.db.PROJECTION,
+            )
+        else:
+            document = await self._db.users.find_one(user_id, PROJECTION)
+
+        document["primary_group"] = {"id": document["primary_group"]}
+
+        return Account(**document)
+
+    async def get_settings(self, query_field: str, user_id: str) -> AccountSettings:
+        """
+        Gets account settings.
+
+        :param query_field: the field to get
+        :param user_id: the user ID
+        :return: the account settings
+        """
+        settings = await get_one_field(self._db.users, query_field, user_id)
+
+        return AccountSettings(**settings)
+
+    async def edit_settings(
+        self, data: EditSettingsSchema, query_field: str, user_id: str
+    ) -> AccountSettings:
+        """
+        Edits account settings.
+
+        :param data: the update to the account settings
+        :param query_field: the field to edit
+        :param user_id: the user ID
+        :return: the account settings
+        """
+        settings_from_db = await get_one_field(self._db.users, query_field, user_id)
+
+        data_dict = data.dict(exclude_unset=True)
+
+        settings = {**settings_from_db, **data_dict}
+
+        await self._db.users.update_one({"_id": user_id}, {"$set": settings})
+
+        return AccountSettings(**settings)
+
+    async def get_keys(
+        self, user_id: str, api_key_projection: Dict[str, bool]
+    ) -> collections.AsyncIterable:
+        """
+        Gets API keys associated with the authenticated user account.
+
+        :param user_id: the user ID
+        :param api_key_projection: the projection for the api keys
+        :return: the api keys
+        """
+        cursor = self._db.keys.find({"user.id": user_id}, api_key_projection)
+
+        return cursor
+
+    async def create_key(self, data: CreateKeysSchema, user_id: str) -> dict:
+        """
+        Create a new API key.
+
+        :param data: the fields to create a new API key
+        :param user_id: the user ID
+        :return: the API key
+        """
+        if isinstance(data.permissions, dict):
+            permissions_dict = data.permissions
+        else:
+            permissions_dict = data.permissions.dict(exclude_unset=True)
```

Comment:
1. Are you sure this is necessary? I would have thought all permission dictionaries used to populate the model would be converted to objects.
2. This is repeated elsewhere. Extract it to a reusable function.

**Number 1 is the preferred change here!**